library/x509_create.c

```diff
@@ -37,44 +37,46 @@ typedef struct {
-    { ADD_STRLEN( "postalCode" ),               MBEDTLS_OID_AT_POSTAL_CODE },
-    { ADD_STRLEN( "dnQualifier" ),              MBEDTLS_OID_AT_DN_QUALIFIER },
-    { ADD_STRLEN( "title" ),                    MBEDTLS_OID_AT_TITLE },
-    { ADD_STRLEN( "surName" ),                  MBEDTLS_OID_AT_SUR_NAME },
-    { ADD_STRLEN( "SN" ),                       MBEDTLS_OID_AT_SUR_NAME },
-    { ADD_STRLEN( "givenName" ),                MBEDTLS_OID_AT_GIVEN_NAME },
-    { ADD_STRLEN( "GN" ),                       MBEDTLS_OID_AT_GIVEN_NAME },
-    { ADD_STRLEN( "initials" ),                 MBEDTLS_OID_AT_INITIALS },
-    { ADD_STRLEN( "pseudonym" ),                MBEDTLS_OID_AT_PSEUDONYM },
-    { ADD_STRLEN( "generationQualifier" ),      MBEDTLS_OID_AT_GENERATION_QUALIFIER },
-    { ADD_STRLEN( "domainComponent" ),          MBEDTLS_OID_DOMAIN_COMPONENT },
-    { ADD_STRLEN( "DC" ),                       MBEDTLS_OID_DOMAIN_COMPONENT },
-    { NULL, 0, NULL }
+    { ADD_STRLEN( "CN" ),                       MBEDTLS_OID_AT_CN,              MBEDTLS_ASN1_UTF8_STRING },
+    { ADD_STRLEN( "commonName" ),               MBEDTLS_OID_AT_CN,              MBEDTLS_ASN1_UTF8_STRING },
+    { ADD_STRLEN( "C" ),                        MBEDTLS_OID_AT_COUNTRY,         MBEDTLS_ASN1_PRINTABLE_STRING },
+    { ADD_STRLEN( "countryName" ),              MBEDTLS_OID_AT_COUNTRY,         MBEDTLS_ASN1_PRINTABLE_STRING },
+    { ADD_STRLEN( "O" ),                        MBEDTLS_OID_AT_ORGANIZATION,    MBEDTLS_ASN1_UTF8_STRING },
+    { ADD_STRLEN( "organizationName" ),         MBEDTLS_OID_AT_ORGANIZATION,    MBEDTLS_ASN1_UTF8_STRING },
+    { ADD_STRLEN( "L" ),                        MBEDTLS_OID_AT_LOCALITY,        MBEDTLS_ASN1_UTF8_STRING },
+    { ADD_STRLEN( "locality" ),                 MBEDTLS_OID_AT_LOCALITY,        MBEDTLS_ASN1_UTF8_STRING },
+    { ADD_STRLEN( "R" ),                        MBEDTLS_OID_PKCS9_EMAIL,        MBEDTLS_ASN1_IA5_STRING},
+    { ADD_STRLEN( "OU" ),                       MBEDTLS_OID_AT_ORG_UNIT,        MBEDTLS_ASN1_UTF8_STRING },
+    { ADD_STRLEN( "organizationalUnitName" ),   MBEDTLS_OID_AT_ORG_UNIT,        MBEDTLS_ASN1_UTF8_STRING },
+    { ADD_STRLEN( "ST" ),                       MBEDTLS_OID_AT_STATE,           MBEDTLS_ASN1_UTF8_STRING },
+    { ADD_STRLEN( "stateOrProvinceName" ),      MBEDTLS_OID_AT_STATE,           MBEDTLS_ASN1_UTF8_STRING },
+    { ADD_STRLEN( "emailAddress" ),             MBEDTLS_OID_PKCS9_EMAIL,        MBEDTLS_ASN1_IA5_STRING },
+    { ADD_STRLEN( "serialNumber" ),             MBEDTLS_OID_AT_SERIAL_NUMBER,   MBEDTLS_ASN1_PRINTABLE_STRING },
+    { ADD_STRLEN( "postalAddress" ),            MBEDTLS_OID_AT_POSTAL_ADDRESS,  MBEDTLS_ASN1_PRINTABLE_STRING },
+    { ADD_STRLEN( "postalCode" ),               MBEDTLS_OID_AT_POSTAL_CODE,     MBEDTLS_ASN1_PRINTABLE_STRING },
+    { ADD_STRLEN( "dnQualifier" ),              MBEDTLS_OID_AT_DN_QUALIFIER,    MBEDTLS_ASN1_PRINTABLE_STRING },
```

Comment:
@thomas-dee Could you explain the varying choices of encodings here?

Reply:
As mentioned in line 45 I tried to default to the types given in RFC3280. For the given types it was not clear for me, so I defaulted to `PrintableString`.

Reply:
@thomas-dee I overlooked the reference to RFC 3280, thanks for pointing that out.

Reply:
We should reference the newer [RFC 5280](https://tools.ietf.org/html/rfc5280#appendix-A.1) here.